synapse/util/ratelimitutils.py

```diff
@@ -86,7 +88,9 @@
 )
 
 
-_rate_limiter_instances: Set["FederationRateLimiter"] = set()
+# This must be a `WeakSet`, otherwise we indirectly hold on to entire `HomeServer`s
+# during trial test runs and leak a lot of memory.
+_rate_limiter_instances: MutableSet["FederationRateLimiter"] = WeakSet()
```

Comment:
Not familiar with weakset. Does this mean that if I do:

```python
x = (1,2)
w = WeakSet()
w.insert(x)
del x
```
Then `len(w) == 0`?

Reply:
Ahh, only certain types of objects can hold weakrefs per https://docs.python.org/3/library/weakref.html

```
>>> x = frozenset("abc")
>>> w.add(x)
>>> len(w)
1
>>> del x
>>> len(w)
0
```

Reply:
That's roughly the idea, yep. The set won't prevent garbage collection of any of its contents and GCed objects will magically disappear from the set.